### ISSUE TITLE
DiagrammedAssertions.assert causes compiler failures when using `new`

### ISSUE THREAD

**Issue by Grundlefleck** [scalatest/scalatest#577]
I have the following spec:

```
import org.scalatest.{DiagrammedAssertions, FunSpec}

class SomeOtherClass(field: Int)
case class MyCaseClass(stringyThing: String, something: SomeOtherClass)

class MacroCompileErrorSpec extends FunSpec with DiagrammedAssertions {
  describe("assert macro") {
    it("should not fail to compile when comparing to a case class created with a field constructed inline") {
      val something = MyCaseClass("a", new SomeOtherClass(1))

      assert(something == MyCaseClass("b", new SomeOtherClass(2)))
    }
  }
}
```

This fails to compile with the following error message:
```
Error:scalac: Error: Unexpected New(TypeTree(class SomeOtherClass)/SomeOtherClass) received in icode.
  Call was genLoad(new SomeOtherClass,package <empty> { class MacroCompileErrorSpec$$anonfun$1$$anonfun$apply$mcV$sp$1 { def MacroCompileErrorSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp { bb=1 } } },REF(class Object))
scala.reflect.internal.FatalError: Unexpected New(TypeTree(class SomeOtherClass)/SomeOtherClass) received in icode.
  Call was genLoad(new SomeOtherClass,package <empty> { class MacroCompileErrorSpec$$anonfun$1$$anonfun$apply$mcV$sp$1 { def MacroCompileErrorSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp { bb=1 } } },REF(class Object))
	at scala.reflect.internal.Reporting$class.abort(Reporting.scala:59)
	at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:16)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:598)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genLoadArguments$1.apply(GenICode.scala:1133)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genLoadArguments$1.apply(GenICode.scala:1131)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:110)
	at scala.collection.immutable.List.foldLeft(List.scala:83)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadArguments(GenICode.scala:1131)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadApply6$1(GenICode.scala:778)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:809)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadValDef$1(GenICode.scala:525)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:533)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genStat(GenICode.scala:181)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:110)
	at scala.collection.immutable.List.foldLeft(List.scala:83)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:915)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadValDef$1(GenICode.scala:525)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:533)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genStat(GenICode.scala:181)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:110)
	at scala.collection.immutable.List.foldLeft(List.scala:83)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:915)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadValDef$1(GenICode.scala:525)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:533)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genStat(GenICode.scala:181)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:110)
	at scala.collection.immutable.List.foldLeft(List.scala:83)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:915)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadValDef$1(GenICode.scala:525)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:533)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genStat(GenICode.scala:181)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase$$anonfun$genStat$1.apply(GenICode.scala:155)
	at scala.collection.LinearSeqOptimized$class.foldLeft(LinearSeqOptimized.scala:110)
	at scala.collection.immutable.List.foldLeft(List.scala:83)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:915)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:916)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:123)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:71)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:148)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:98)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:71)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:89)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.gen(GenICode.scala:67)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.apply(GenICode.scala:63)
	at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:410)
	at scala.tools.nsc.Global$GlobalPhase$$anonfun$run$1.apply(Global.scala:377)
	at scala.tools.nsc.Global$GlobalPhase$$anonfun$run$1.apply(Global.scala:377)
	at scala.collection.Iterator$class.foreach(Iterator.scala:743)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1177)
	at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:377)
	at scala.tools.nsc.backend.icode.GenICode$ICodePhase.run(GenICode.scala:55)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1557)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1542)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1537)
	at scala.tools.nsc.Global$Run.compile(Global.scala:1644)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:126)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:102)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:102)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:48)
	at sbt.compiler.AnalyzingCompiler.compile(AnalyzingCompiler.scala:41)
	at org.jetbrains.jps.incremental.scala.local.IdeaIncrementalCompiler.compile(IdeaIncrementalCompiler.scala:29)
	at org.jetbrains.jps.incremental.scala.local.LocalServer.compile(LocalServer.scala:26)
	at org.jetbrains.jps.incremental.scala.remote.Main$.make(Main.scala:62)
	at org.jetbrains.jps.incremental.scala.remote.Main$.nailMain(Main.scala:20)
	at org.jetbrains.jps.incremental.scala.remote.Main.nailMain(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.martiansoftware.nailgun.NGSession.run(NGSession.java:319)
```

This occurs both in IntelliJ and sbt, on ScalaTest `2.2.4`. I have constructed an example repo containing a SSCCE here: https://github.com/Grundlefleck/scalatest-constructor-in-assert-macro 

Is this a bug or a known limitation?

Many thanks.

**Comment by cheeseng** [comment]
@Grundlefleck Actually the fix for this problem is already available: 

https://github.com/scalatest/scalatest/commit/a6674aa97664f4449abe5ece009a694c22909a14

For some reason, the fix is not included in the branch that was used to build 2.2.4, I'll check with @bvenners to see if we can include it in 2.2.5.

Thanks.

**Comment by bvenners** [comment]
We'll do a 2.2.5 to release this bug fix.

**Comment by cheeseng** [comment]
@Grundlefleck I just make a 2.2.5 release that includes the fix, can you try to see if it fixes your problem?

**Comment by Grundlefleck** [comment]
Hi @cheeseng,

Thanks for the release fix. It does address the case I used in my example, but unfortunately, not the real case I encountered. While it is now happy constructing a Scala class inline, it appears to have problems calling static methods from Java. In particular, a method from the JodaTime library.

To try to help hone in on the issue, I've updated the failing test case here:

https://github.com/Grundlefleck/scalatest-constructor-in-assert-macro/blob/master/src/test/scala/MacroCompileErrorSpec.scala#L29

**Comment by cheeseng** [comment]
@Grundlefleck bummer, I think this Java static methods problem is new, I'll investigate it with fresh mind when starting tomorrow morning.

**Comment by Grundlefleck** [comment]
@cheeseng Thanks! We do have a workaround of this issue. extract a local val instead of constructing inline, so it's not the end of the world.

Thanks!

**Comment by cheeseng** [comment]
@Grundlefleck i submitted a fix here: 

https://github.com/scalatest/scalatest/pull/583

it is against 2.3.x branch through, I think it should be easy for us to cherry pick it into 2.2 branch, would you mind helping us to verify that it has fixed your problem?

**Comment by Grundlefleck** [comment]
Hi,

I build 2.3.0-SNAP2 locally and the test now compiles, good stuff! Don't worry about cherry picking back on to the 2.2 branch on our behalf, we're happy to wait for the fix to appear in 2.3.

Thanks!
